tests/test_templates_broadcast.py

```diff
@@ -0,0 +1,267 @@
+    def test_correct_parameters_in_queue(self, structure, gate, parameters):
+        """Tests that gate queue has correct parameters."""
+
+        with qml.utils.OperationRecorder() as rec:
+            broadcast(block=gate, structure=structure, wires=range(3), parameters=parameters)
+
+        for target_par, g in zip(parameters, rec.queue):
+            assert g.parameters == target_par
+
+    @pytest.mark.parametrize("structure, pars1, pars2, gate", [("double", [[], []], None, CNOT),
+                                                               ("double", [1, 2], [[1], [2]], CRX)])
+    def test_double_correct_queue_for_different_parameter_formats(self, structure, pars1, pars2, gate):
+        """Tests that specific parameter formats have the same output."""
+
+        with qml.utils.OperationRecorder() as rec1:
+            broadcast(block=gate, structure=structure, wires=range(4), parameters=pars1)
+
+        with qml.utils.OperationRecorder() as rec2:
+            broadcast(block=gate, structure=structure, wires=range(4), parameters=pars2)
+
+        for g1, g2 in zip(rec1.queue, rec2.queue):
+            assert g1.parameters == g2.parameters
+
+    @pytest.mark.parametrize("structure, parameters, unitary, target", TARGET_OUTPUTS)
+    def test_prepares_correct_state(self, structure, parameters, unitary, target):
+        """Tests the state produced by different unitaries."""
+
+        @qml.qnode(dev_4_qubits)
+        def circuit():
+            for w in range(4):
+                qml.PauliX(wires=w)
+            broadcast(block=unitary, structure=structure, wires=range(4), parameters=parameters)
+            return [qml.expval(qml.PauliZ(wires=w)) for w in range(4)]
+
+        res = circuit()
+        assert np.allclose(res, target)
+
+    @pytest.mark.parametrize("structure, parameters, unitary, target, observable", CV_TARGET_OUTPUTS)
+    def test_prepares_correct_state_cv(self, structure, parameters, unitary, target, observable):
+        """Tests the state produced by different unitaries."""
+
+        @qml.qnode(dev_4_qumodes)
+        def circuit():
+            for w in range(4):
+                Displacement(1, 0, wires=w)
+            broadcast(block=unitary, structure=structure, wires=range(4), parameters=parameters)
+            return [qml.expval(observable(wires=w)) for w in range(4)]
+
+        res = circuit()
+        assert np.allclose(res, target)
+
+    @pytest.mark.parametrize("parameters, n_wires", [(np.array([0]), 2),
+                                                     ([0, 0, 0, 1, 0], 3)])
+    def test_throws_error_when_mismatch_params_wires(self, parameters, n_wires):
+        """Tests that error thrown when 'parameters' does not contain one set
+           of parameters for each wire."""
+
+        dev = qml.device('default.qubit', wires=n_wires)
+
+        @qml.qnode(dev)
+        def circuit():
+            broadcast(block=RX, wires=range(n_wires), parameters=parameters)
```

Comment:
```suggestion
            broadcast(block=RX, `'structure'=single`, wires=range(n_wires), parameters=parameters)
```

Reply:
Since no default is defined now

Reply:
Agree with @antalszava, would be best from user's perspective if "single" was the default of `broadcast` (so they'd never have to actually write it!)

Reply:
Yes, that would be really nice @co9olguy as it would provide extra convenience to the user :)

There are certain cases, however, when this would lead to errors. These are the ones where a user defines a multi-qubit template but leaves the default `structure="single"` option. In these cases, we are not too sure how we could deduce on how many qubits the template would act on and would have to rely on the user-specified value.

So the question is whether to:
* cover most cases with a user convenience (by having `"single"` as default), but in certain cases this would result in errors related to wires
* require the user to always pass this as an argument

There was a more detailed discussion in this PR on this issue, though it became quite hidden here on GitHub (might need to wait a bit until the messages open in the thread): https://github.com/XanaduAI/pennylane/pull/515#discussion_r382243785

An alternative way could actually be having a `try-except` block, and catch the first time such an error would be raised. Though this might not be too elegant of a solution.

Reply:
Might be enough to just raise an error if the default is used with a template of a size that doesn't naturally make sense

Reply:
Yes, that's a good idea! The thing is, that for templates it's not so easy how to deduce the size (but maybe we have missed something in our discussion). For PL `Operations` it is contained in `num_wires`, but user-defined templates are simple callable functions.

One way would be to modify the `Recorder` class. Since in its `queue` the operation instances (including the `wires` they are called with) are already present, the overall `wires` could be extracted for the template as well. A question I'm unsure how to solve is where to store this information such that it can be extracted easily (the `template` decorater which uses the `Recorder` class merely returns `recorder.queue`, a list of operations). All in all, this approach would have a couple of edge cases in line and might actually require more thinking.

Reply:
Could the template decorator be modified to add attributes to the python function? e.g.,

```python
@template
def ansatz(weights):
    qml.template.StronglyEntanglingLayers(weights, wires=[0, 2, 3])
    qml.Hadamard(wires=4)

>>> ansatz.num_wires
4
>>> anstaz.wires
[0, 2, 3, 4]
```

Reply:
That would be cool @josh146 .

Thanks for summarising all the previous discussion so well @antalszava. I propose we run with no default for now and reassess this, since a) it will not be feature breaking and b) it is really a lot more complex.

Reply:
> That would be cool @josh146 .

I believe it should be possible, we can create an issue for it for a future PR.